File: Java/ProtoDebugger/src/protodebugger/model/descriptors/EnumFieldDescriptorContainer.java
package protodebugger.model.descriptors;

import org.eclipse.swt.SWT;
import org.eclipse.swt.widgets.Combo;
import org.eclipse.swt.widgets.Composite;
import org.eclipse.swt.widgets.Widget;

import com.google.protobuf.Descriptors;
import com.google.protobuf.Descriptors.EnumValueDescriptor;
import com.google.protobuf.GeneratedMessage.Builder;


public class EnumFieldDescriptorContainer extends FieldDescriptorContainer {
	private Combo comboField;
	
	public EnumFieldDescriptorContainer(Descriptors.FieldDescriptor field)
	{
		super(field);
	}
	
	@Override
	public Object getValue() {
		if(value != null)
			return value.toString();
		else if(defaultValue != null)
			return ((EnumValueDescriptor)defaultValue).getName();
		else 
			return field.getEnumType().getValues().get(0).toString();
	}

	@Override
	public boolean buildMsg(Builder<?> build) {
		if(field.isOptional() && comboField.getText().equals(""))
			return false;
		if(field.isRepeated())
			build.addRepeatedField(field, field.getEnumType().findValueByName(comboField.getText()));
		else
			build.setField(field, field.getEnumType().findValueByName(comboField.getText()));
		return true;
	}

	@Override
	public void setValue(Object value) {
		if(value instanceof String)
		{
			for(EnumValueDescriptor evd : field.getEnumType().getValues())
			{
				if(evd.getName().equals(value))
				{
					value = evd;
					break;
				}
			}
		}
	}

	@Override
	public String toString() 
	{
		return "EnumField name = " + name;
	}

	@Override
	public Widget getWidget(Composite parent) 
	{
		if(comboField == null)
		{
			comboField = new Combo(parent, SWT.READ_ONLY);
			comboField.setItems(getValues());
			if(defaultValue != null)
				comboField.select(((EnumValueDescriptor)
						defaultValue).getIndex());
		}
		else if(comboField.getParent() != parent)
		{
			comboField.setParent(parent);
		}
		return comboField;
	}
	
	public String[] getValues()
	{
		int arraySize = field.getEnumType().getValues().size();
		if(field.isOptional())
			arraySize += 1;
		String[] ret = new String[arraySize];
		for(EnumValueDescriptor evd : field.getEnumType().getValues())
			ret[evd.getIndex()] = evd.getName();
		if(field.isOptional())
			ret[arraySize-1] = "";
		return ret;
	}

	@Override
	public Composite getParent() {
		if(comboField != null)
			return comboField.getParent();
		return null;
	}

}

File: CPP/ProtoDebugger/src/protodebugger/model/descriptors/EnumFieldDescriptorContainer.cpp
#include "EnumFieldDescriptorContainer.h"
#include <sstream>

using namespace std;
using namespace google;
using namespace protobuf;

QWidget * EnumFieldDescriptorContainer::getWidget(QWidget * parent)
{
    if(m_comboBox == NULL)
    {
        m_comboBox = new QComboBox(parent);

        if (m_type == FieldDescriptor::CPPTYPE_ENUM)
        {
            m_comboBox->addItems(getValues());

            if(m_defaultValue != NULL)
            {
                EnumValueDescriptor * value = (EnumValueDescriptor *) m_defaultValue;
                m_comboBox->setCurrentIndex(value->index());
            }
        }
    }
    else if (m_comboBox->parentWidget() != parent)
    {
        m_comboBox->setParent(parent);
    }

    return m_comboBox;
}

QWidget * EnumFieldDescriptorContainer::getParent()
{
    if(m_comboBox != NULL)
    {
       return m_comboBox->parentWidget();
    }

    return NULL;
}

Object * EnumFieldDescriptorContainer::getValue()
{
    stringstream ss;
    if(m_value != NULL)
    {
        if (m_type == FieldDescriptor::CPPTYPE_ENUM)
        {
            ss << m_value;
            return (Object *) ss.str().c_str();
        }
     }
    else if (m_defaultValue != NULL)
    {
        if (m_type == FieldDescriptor::CPPTYPE_ENUM)
        {
            EnumValueDescriptor * value = (EnumValueDescriptor *) m_defaultValue;
            return (Object *) value->name().c_str();
        }
    }
    return (Object *) m_field->enum_type()->value(0)->name().c_str();
}

void EnumFieldDescriptorContainer::setValue(Object * value)
{
    char * result = (char *) value;
    string check = result;
    if(check.c_str() != NULL)
    {
        for (int i = 0; i < m_field->enum_type()->value_count(); ++i)
        {
            if(m_field->enum_type()->name().compare(check) == 0)
            {
                m_value = (Object *)m_field->enum_type();
                return;
            }
        }
    }
}

QString EnumFieldDescriptorContainer::toString()
{
    string str = "EnumField name = " + this->m_name;
    return QString(str.c_str());
}

QStringList EnumFieldDescriptorContainer::getValues()
{
    int arraySize = m_field->enum_type()->value_count();
    if (m_field->is_optional())
    {
        arraySize += 1;
    }
    QStringList result;

    for (int i = 0; i < arraySize; ++i)
    {
        result << m_field->enum_type()->value(i)->name().c_str();
    }
    if(m_field->is_optional())
    {
        result << "";
    }
    return result;
}

bool EnumFieldDescriptorContainer::buildMsg(Message *msg)
{
    if( m_field->is_optional() && m_comboBox->currentText().compare("") == 0)
    {
        return false;
    }
    if(m_field->is_repeated())
    {
        int index = 0;
        QStringList check = this->getValues();

        for(int i = 0; i < m_field->enum_type()->value_count(); ++i)
        {
            if( m_field->enum_type()->value(i)->name().compare(check.at(i).toStdString()) == 0)
            {
                index = i;
                break;
            }
        }
        msg->GetReflection()->AddEnum(msg, m_field, m_field->enum_type()->value(index));
    }
    else
    {
        int index = 0;
        QStringList check = this->getValues();

        for(int i = 0; i < m_field->enum_type()->value_count(); ++i)
        {
            if( m_field->enum_type()->value(i)->name().compare(check.at(i).toStdString()) == 0)
            {
                index = i;
                break;
            }
        }
        msg->GetReflection()->SetEnum(msg, m_field, m_field->enum_type()->value(index));
    }
    return true;
}
